src/library/attr-groups.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * attr-groups.c - Linked list as a temporary memory storage
+ * for trust database data
+ * Copyright (c) 2016,2020 Red Hat Inc., Durham, North Carolina.
+ * All Rights Reserved.
+ *
+ * This software may be freely redistributed and/or modified under the
+ * terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2, or (at your option) any
+ * later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING. If not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor
+ * Boston, MA 02110-1335, USA.
+ *
+ * Authors:
+ *    Radovan Sroka <rsroka@redhat.com>
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "attr-groups.h"
+#include "message.h"
+
+#define RESIZE_BY 2
+#define DEFAULT_CAPACITY 100
+
+attr_groups_t groups;
+
+static int strcmp_cb(void * a, void * b)
+{
+	return strcmp(((avl_str_data_t *)a)->str, ((avl_str_data_t *)b)->str);
+}
+
+static int intcmp_cb(void * a, void * b)
+{
+	return ((avl_int_data_t *)a)->num - ((avl_int_data_t *)b)->num;
+}
+
+static int strncmp_cb(void * a, void * b)
+{
+	return strncmp( ((avl_str_data_t *)a)->str,
+			((avl_str_data_t *)b)->str,
+			((avl_str_data_t *)a)->len)
+		? 0 : -1;
```

Comment:
This is a rather terse way of writing the code. At first glance its not obvious what its doing. I'd rather see clarity of purpose by holding the return code as a variable and doing if/else. And a comment about what is returned and its meaning would also be helpful. For example, why is -1 returned? why not 1?

Reply:
I can add there comment.. -1 break the avl tree traverse recursion. When strncmp params are equal (for n characters) it returns 0 then we need to return -1 to stop recursion from working.

Reply:
Avl tree traverse call callback on each node and it sums all return values and it returns the sum.
So when I return 1 it just sums how many strncmp returned match. With -1 I can break the recursion with first match.